overlord/snapstate/handlers.go

```diff
@@ -1148,36 +1148,39 @@ func (m *SnapManager) doLinkSnap(t *state.Task, _ *tomb.Tomb) (err error) {
 func maybeRestart(t *state.Task, info *snap.Info) {
 	st := t.State()
 
-	if release.OnClassic {
-		// ignore error here as we have no way to return to caller
-		snapdSnapInstalled, _ := isInstalled(st, "snapd")
-		if (info.GetType() == snap.TypeOS && !snapdSnapInstalled) ||
-			info.GetType() == snap.TypeSnapd {
-			t.Logf("Requested daemon restart.")
-			st.RequestRestart(state.RestartDaemon)
-		}
+	model, err := ModelFromTask(t)
+	if err != nil {
+		logger.Noticef("cannot get model assertion: %v", model)
 		return
 	}
 
-	// On a core system we may need a full reboot if
-	// core/base or the kernel changes.
-	if boot.ChangeRequiresReboot(info) {
-		t.Logf("Requested system restart.")
-		st.RequestRestart(state.RestartSystem)
+	typ := info.GetType()
+	bp, applicable := boot.Lookup(info, typ, model, release.OnClassic)
+	if applicable {
+		if bp.ChangeRequiresReboot() {
+			t.Logf("Requested system restart.")
+			st.RequestRestart(state.RestartSystem)
+		}
 		return
 	}
 
-	// On core systems that use a base snap we need to restart
-	// snapd when the snapd snap changes.
-	model, err := ModelFromTask(t)
-	if err != nil {
-		logger.Noticef("cannot get model assertion: %v", model)
+	if !((release.OnClassic && typ == snap.TypeOS) || typ == snap.TypeSnapd) {
```

Comment:
I think this should be a cleanly named helper:

```
func RequiresDaemonRestart(onClassic bool, snapType snap.Type) bool {
    return (onClassic && typ == snap.TypeOS) || snapType == snap.TypeSnapd
}
```

Then it is clear that it should also have some input from exec stack (we don't need to restart anything when reexec is not in use).

Then having that you can do:
```
if !RequiresDaemonRestart(release.OnClassic, typ) {
    // not interesting
   return
}
```

and this is more readable IMO as to the intent.